src/diffusers/models/unet_2d_blocks.py

```diff
@@ -432,8 +432,8 @@ def __init__(
                 attentions.append(
                     Attention(
                         in_channels,
-                        heads=in_channels // attn_num_head_channels if attn_num_head_channels is not None else 1,
-                        dim_head=attn_num_head_channels if attn_num_head_channels is not None else in_channels,
+                        heads=in_channels // attn_num_heads if attn_num_heads is not None else 1,
+                        dim_head=attn_num_heads if attn_num_heads is not None else in_channels,
```

Comment:
my head is :exploding_head: a little bit now reading this PR so I could be wrong 

but from what I understand the `attn_num_head_channels` has been passed correctly as `dim_head` for all attention classes except for `Transformer2D` and this change will cause unexpected behavior when we use the new `num_attention_heads` argument from the `UNet2DConditionModel` for other attention class 

Taking `AttnDownBlock2D` as example here

using `attention_head_dim` argument will return expected result
```python
from diffusers import UNet2DConditionModel
down_block_types = ("AttnDownBlock2D",)
up_block_types = ("AttnUpBlock2D",)
unet = UNet2DConditionModel(
    attention_head_dim = 16,
    block_out_channels = (320,),
    down_block_types = down_block_types,
    up_block_types = up_block_types)
# this prints 20
unet.down_blocks[0].attentions[0].heads
```
using `num_attention_heads` argument will return the wrong results
```python
down_block_types = ("AttnDownBlock2D",)
up_block_types = ("AttnUpBlock2D",)
unet = UNet2DConditionModel(
    num_attention_heads = 16,
    block_out_channels = (320,),
    down_block_types = down_block_types,
    up_block_types = up_block_types)
# this prints 20, which is wrong? should be 16
unet.down_blocks[0].attentions[0].heads
```

maybe we need to path both `num_attention_heads` and `attention_head_dim` to the blocks?

Reply:
You're 100% right great catch! I think there was a double incorrect naming correcting itself here haha

Reply:
Shouldn't I expect any mapping between what the user specifies as `num_attention_heads` and `attention_head_dim` and how it's handled internally?

To put this perspective, 

```python
from diffusers import UNet2DConditionModel

down_block_types = ("AttnDownBlock2D",)
up_block_types = ("AttnUpBlock2D",)
unet = UNet2DConditionModel(
    num_attention_heads = 16,
    block_out_channels = (320,),
    down_block_types = down_block_types,
    up_block_types = up_block_types
)

# this prints 40
unet.down_blocks[0].attentions[0].heads
```

and

```python
down_block_types = ("AttnDownBlock2D",)
up_block_types = ("AttnUpBlock2D",)
unet = UNet2DConditionModel(
    attention_head_dim = 16,
    block_out_channels = (320,),
    down_block_types = down_block_types,
    up_block_types = up_block_types)
# this prints 20
unet.down_blocks[0].attentions[0].heads
```

Are all these expected? Like shouldn't `unet.down_blocks[0].attentions[0].heads` print 16 unless I am missing out on something?

Reply:
Just in case you don't miss it, pinging @patrickvonplaten (apologies in advance).

Reply:
Hmm immediately, I don't have any idea to mitigate that faithfully, though.

Reply:
@pcuenca i thought both `AttnDownBlock2D` and `CrossAttnDownBlock2D` need this argument no? anything with the `Attention` class